PurchasesCoreSwift/Attribution/AttributionTypeFactory.swift

```diff
@@ -0,0 +1,72 @@
+//
+//  AttributionTypeFactory.swift
+//  PurchasesCoreSwift
+//
+//  Created by Juanpe Catalán on 9/7/21.
+//  Copyright © 2021 Purchases. All rights reserved.
+//
+
+import Foundation
+
+// TODO(Post-migration): switch this back to internal the class and all these protocols and properties.
+
+public typealias AttributionDetailsBlock = ([String: Any]?, Error?) -> Void
+
+@objc public enum FakeTrackingManagerAuthorizationStatus: Int {
+    case notDetermined = 0
+    case restricted
+    case denied
+    case authorized
+}
+
+@objc public protocol FakeAdClient: AnyObject {
+    static func sharedClient() -> FakeAdClient
+    @objc(requestAttributionDetailsWithBlock:)
+    func requestAttributionDetails(_ completionHandler: AttributionDetailsBlock)
+}
+
+@objc public protocol FakeASIdentifierManager: AnyObject {
+    static func sharedManager() -> FakeASIdentifierManager
+}
+
+@objc public protocol FakeTrackingManager: AnyObject {
+    static func trackingAuthorizationStatus() -> Int
+}
+
+@objc(RCAttributionTypeFactory)
+open class AttributionTypeFactory: NSObject {
+    let mangledIdentifierClassName = "NFVqragvsvreZnantre"
+    let mangledIdentifierPropertyName = "nqiregvfvatVqragvsvre"
+    let mangledAuthStatusPropertyName = "genpxvatNhgubevmngvbaFgnghf"
+    let mangledTrackingClassName = "NGGenpxvatZnantre"
+
+    @objc open func adClientClass() -> FakeAdClient.Type? {
+        NSClassFromString("ADClient") as? FakeAdClient.Type
+    }
+
+    @objc open func atTrackingClass() -> FakeTrackingManager.Type? {
+        // We need to do this mangling to avoid Kid apps being rejected for getting idfa.
+        // It looks like during the app review process Apple does some string matching looking for
+        // functions in ATTrackingTransparency. We apply rot13 on these functions and classes names
+        // so that Apple can't find them during the review, but we can still access them on runtime.
+        let className = mangledTrackingClassName.rot13()
+        return NSClassFromString(className) as? FakeTrackingManager.Type
```

Comment:
This is the bit that won't work in Swift - even though the class is correctly obtained in `NSClassFromString`, it won't be castable to `FakeTrackingManager`, since the class doesn't explicitly conform to the protocol. 
Swift doesn't have reflection, so we can't call the method by name either. 

I'm honestly not sure what the best fix for this is, would love suggestions

Reply:
Here are some steps to reproduce: 

Here are some testing steps to help figure this out: 

In an app that integrates the RevenueCat SDK (you can use the one in `IntegrationTests/PurchaseTester`, you'll have to run `pod install` and update the signing team to your own).
- in Build Phases, in Link Binary with Libraries, add `AdServices` and `AppTrackingTransparency`. 
- in the app, on app launch, add the following code. This will trigger a tracking request, make sure you select Allow. If the tracking request doesn't show up, make sure your device doesn't have tracking disabled or enable it in `Settings -> Privacy -> Tracking -> Allow Apps to Request to Track`
```swift
ATTrackingManager.requestTrackingAuthorization { status in
    Purchases.automaticAppleSearchAdsAttributionCollection = true
    Purchases.configure(withAPIKey: <your_public_sdk_key_here>)
    Purchases.logLevel = .debug
}
```

This will trigger: a modal to track (again, make sure you accept), then a call to these methods to figure out whether the frameworks are linked in the app, instantiate their classes and call their corresponding methods. 
Make sure you set a breakpoint here to see if the class can be obtained with `NSClassFromString(className)`, and then casted with `as? FakeTrackingManager.Type`.

Reply:
Ohh, yeah, I can see the problem 🤔 I've checked the objc runtime API but I can't find a valid solution for that...Firstly, I thought that maybe we could replace the implementation of the fake method with the original one by swizzling but I couldn't find the way.

I've checked also the idea of using the `unsafeBitCast` method to create an empty class with the same size as `ATTrackingManager` and replace the implementation of the methods instead of using a `FakeProtocol` but I was facing an error as well :/

It seems that the solution provided by @taquitos could be useful, right?

BTW, I think there is another approach but it's harder for the developers. The idea is inversing the control, I mean, we only need to know if the tracking is authorized or not, so we could use a closure or a delegate method to ask to the developer code the current status. The code would look like this:

```swift
ATTrackingManager.requestTrackingAuthorization { status in
    Purchases.trackingAuthorizationStatusProvider = <Closure or delegate here>
    Purchases.automaticAppleSearchAdsAttributionCollection = true
    Purchases.configure(withAPIKey: <your_public_sdk_key_here>)
    Purchases.logLevel = .debug
}
```

The problem with this approach is the migration would change the logic and the API and I read in the Swift migration plan that you only want to do a translation from ObjC to Swift.

Reply:
Hi again, I think I found a possible way to do that.

As [Apple's doc](https://developer.apple.com/documentation/swift/anyclass) says if you use`AnyClass`: 
> _"all known @objc class methods and properties are available as implicitly unwrapped optional methods and properties"_

So, if we use a `class` instead of a `protocol` to define our `FaceTrackingManager` and we add all the methods that we need. Imagine this definition:

```swift
class FakeTrackingManager {
    @objc class func trackingAuthorizationStatus() -> Int {
        -1
    }
    
    @objc class func requestTrackingAuthorization(completionHandler completion: @escaping (Int) -> Void) {
        // do nothing
    }
}
```

We could create a "`proxy`", something like `TrackingProxy` that perform the methods that we need. For example:
```swift
class TrackingProxy {
    static func trackingAuthorizationStatus(_ trackingClass: AnyClass) -> Int {
        trackingClass.trackingAuthorizationStatus()
    }
    
    static func requestTrackingAuthorization(_ trackingClass: AnyClass, completion: @escaping (Int) -> Void) {
        trackingClass.requestTrackingAuthorization(completionHandler: completion)
    }
}
```

Then, we only need to initialize the class type, `ATTrackingManager` from the mangled string, and pass it as param in the proxy methods:

```swift
let classType: AnyClass = NSClassFromString(mangledTrackingClassName.rot13()) ?? FakeTrackingManager.self
        
TrackingProxy.requestTrackingAuthorization(classType, completion: { status in
                                                        print(status)
                                                   })
```

**Sample app**
<img width="200px" src="https://user-images.githubusercontent.com/1409041/125303511-2775d800-e32d-11eb-9e0e-d8f904ab52f9.png"/>

If we tap on **"Ask App Not to Track"** and execute this code:
```swift
let classType: AnyClass = NSClassFromString(mangledTrackingClassName.rot13()) ?? FakeTrackingManager.self
print(TrackingProxy.trackingAuthorizationStatus(classType)) // 2
```
 the status is: `denied`

WDYT? Do you think this could be useful for us?

Reply:
This looks even smarter than what I suggested!

Reply:
Great! if it looks good for you, I'll make a proposal using this approach, ok?

Reply:
Yeah, it looks great to me 🤩